lib/matplotlib/backends/backend_mixed.py

```diff
@@ -62,27 +62,16 @@ def __init__(self, figure, width, height, dpi, vector_renderer,
 
         self._bbox_inches_restore = bbox_inches_restore
 
-        self._set_current_renderer(vector_renderer)
-
-    _methods = """
-        close_group draw_image draw_markers draw_path
-        draw_path_collection draw_quad_mesh draw_tex draw_text
-        finalize flipy get_canvas_width_height get_image_magnification
-        get_texmanager get_text_width_height_descent new_gc open_group
-        option_image_nocomposite points_to_pixels strip_math
-        start_filter stop_filter draw_gouraud_triangle
-        draw_gouraud_triangles option_scale_image
-        _text2path _get_text_path_transform height width
-        """.split()
-
-    def _set_current_renderer(self, renderer):
-        self._renderer = renderer
-
-        for method in self._methods:
-            if hasattr(renderer, method):
-                setattr(self, method, getattr(renderer, method))
-        renderer.start_rasterizing = self.start_rasterizing
-        renderer.stop_rasterizing = self.stop_rasterizing
+        self._renderer = vector_renderer
+
+    def __getattr__(self, attr):
+        # Proxy everything that hasn't been overridden to the base
+        # renderer. Things that *are* overridden can call methods
+        # on self._renderer directly, but must not cache/store
+        # methods (because things like RendererAgg change their
+        # methods on the fly in order to optimise proxying down
+        # to the underlying C implementation).
+        return getattr(self._renderer, attr)
```

Comment:
I would prefer explicitly listing the functions that get forwarded to the renderer, whether by checking `attr in _methods` in `__getattr__`, or by writing a bunch of forwarding functions
```
for _attr in _methods:
    locals()[_attr] = lambda self, *args, **kwargs: getattr(self._renderer, _attr)(*args, **kwargs)
```
(in spirit; add a call to functools.wraps, etc.)

The reason is that the Renderer API is already quite large, and while implementing https://github.com/anntzer/mplcairo I found out that there are some methods that are not documented but are actually needed for a fully functional renderer backend; making the list explicit should hopefully avoid letting that list further grow out of control.

Reply:
I can see your point of view here @anntzer, but I disagree. In this case we are implementing a renderer *proxy*. A proxy, IMO, should forward *everything* on to the thing it is proxying. It is the thing its proxying's responsibility to define what is and isn't allowed.

In that regard, the list of methods should be documented in the base, not the proxy. Indeed, they *are* partially documented at https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/backend_bases.py#L135-L145, with the remainder being only documented as stub methods.

The simplest argument against explicitly naming all methods is this:

 > if we update the base renderer to have new capabilities, we should not have to update the proxy renderer

But also:

 > if a subclass of a base renderer implements it own methods, the proxy renderer should honour (i.e. proxy) them

Reply:
I agree with not explicitly listing the methods here.

Reply:
Away from serious reviewing for the Christmas period but the complete proxying approach of this PR is fine too even though I think (per the comment below) that a more explicit approach is better.

Reply:
I also agree that we should proxy out everything.